### ISSUE TITLE
SURFACE_2D_FORMAT_X8R8G8B8_Z8R8G8B8 Is unimpletemented 

### ISSUE THREAD

**Issue by Triticum0** [xemu-project/xemu#578]
### Title
https://xemu.app/titles/41560005/#Blade-II

### Bug Description

When trying to load the in to the intro the game crashes due to a missing surface format 0x6.

### Expected Behavior

The intro should work properly.

### xemu Version

Version: 0.6.2
Branch: master
Commit: 69ceec4
Date: Wed Nov 24 02:59:42 UTC 2021
Last-known Working: Unknown

### System Information

Field | Value
-- | --
OS | Ubuntu 20.04.3 LTS
CPU | Intel(R) Pentium(R) CPU 3825U @ 1.90GHz
Graphics Device | Mesa Intel(R) HD Graphics (BDW GT1)
Graphics Driver | Drivers: 4.6 (Core Profile) Mesa 21.0.3

**Comment by abaire** [comment]
Surface format 6 looks like it's [`X8R8G8B8_Z8R8G8B8`](https://gitlab.freedesktop.org/xorg/driver/xf86-video-nouveau/-/blob/master/src/hwdefs/nv01_2d.xml.h#L260). From the dump that @Triticum0 provided it looks like it's doing a mix of 2D and 3D ops, presumably rendering something 3D into the zbuffer and then blending a 2D image against it.

[blade_2_crash.txt](https://github.com/mborgerson/xemu/files/7635478/blade_2_crash.txt)

**Comment by Triticum0** [comment]
Off Topic, but is it useful for me to dump the debug output for game which currently have unexplained crashing/freezing? #376

**Comment by abaire** [comment]
It might be, but I think it's likely only going to be useful for places where it's clearly graphics related (e.g., asserts in `pgraph.c` or one of the nv2a-related files, crashes in your video card driver) and probably only for the crash cases. 

For the freezes I think it's likely to be much more challenging; if xemu is still responsive but the emulated program is not, I think there's a good chance that the emulated machine has crashed. E.g., Return to Castle Wolfenstein was a subtle combination of the emulation of a CPU instruction and the way their code was using the results that led to an invalid memory read that halted the emulated machine.

I think the most useful things for debugging freeze cases would be:
1. A clear description of how to get from a cold startup to the frozen state.
2. Maybe a save file and/or xemu state file just before the freeze to cut down on the amount of time needed to get to the repro case. (Esp. for games that freeze only after playing through multiple levels).

It also might be interesting for xemu to log faults in the emulated machine since that'll give us an idea of what's going on (e.g., knowing that RtCW was crashing due to an access violation and not stuck in a busy wait would've helped point out where to start debugging). @mborgerson I'm not sure if something like this already exists as part of QEMU/xemu (admittedly I haven't always paid enough attention to the console output so maybe it's already doing this)?

**Comment by abaire** [comment]
For posterity, `Z8R8G8B8` just means it forces alpha to zero, it has nothing to do with z-buffering and the mix of 2d and 3d ops in the logs is a red herring.

**Comment by abaire** [comment]
> It might be, but I think it's likely only going to be useful for places where it's clearly graphics related (e.g., asserts in `pgraph.c` or one of the nv2a-related files, crashes in your video card driver) and probably only for the crash cases.
> 
> For the freezes I think it's likely to be much more challenging; if xemu is still responsive but the emulated program is not, I think there's a good chance that the emulated machine has crashed. E.g., Return to Castle Wolfenstein was a subtle combination of the emulation of a CPU instruction and the way their code was using the results that led to an invalid memory read that halted the emulated machine.
> 
> I think the most useful things for debugging freeze cases would be:
> 
> 1. A clear description of how to get from a cold startup to the frozen state.
> 2. Maybe a save file and/or xemu state file just before the freeze to cut down on the amount of time needed to get to the repro case. (Esp. for games that freeze only after playing through multiple levels).
> 
> It also might be interesting for xemu to log faults in the emulated machine since that'll give us an idea of what's going on (e.g., knowing that RtCW was crashing due to an access violation and not stuck in a busy wait would've helped point out where to start debugging). @mborgerson I'm not sure if something like this already exists as part of QEMU/xemu (admittedly I haven't always paid enough attention to the console output so maybe it's already doing this)?

FYI @Triticum0 
https://github.com/abaire/xemu/commit/bb5e9c01b9daae4b6789fd1c330c6d073ab04359 adds very basic logging that at least seems to catch the Brother In Arms Road To Hill 30 freeze. I might poke around a bit more and see if it's worth a PR to help debug the rest of the freeze issues.

**Comment by Triticum0** [comment]
I tested Brother In Arms: Earned in Blood and it gave this assertion: 
MMU fault: ExceptionIndex: EXCP0E ErrorCode: 0 ReturnAddr: 7F897B95F9F3 EIP: 33409D
MMU fault: ExceptionIndex: EXCP0E ErrorCode: 0 ReturnAddr: 7F897B979CB5 EIP: 80049FE1

Using your build will get back to you if i find anything else

**Comment by abaire** [comment]
Awesome! Do you mind tracking any of those `MMU fault` messages on #376 ? I don't think those freezes are related to this bug.

**Comment by Triticum0** [comment]
Should I make a separate issue for the MMU fault instead. The whole point of #376 is it unknown why it crashes once it a known issue i think it best to separate the issue

**Comment by abaire** [comment]
> Should I make a separate issue for the MMU fault instead. The whole point of #376 is it unknown why it crashes once it a known issue i think it best to separate the issue

I think that'd be fine, but I'd make an issue for Earned In Blood specifically rather than focus on the MMU fault. That fault is arguably more of a symptom of something else being in a bad state than the direct cause. It's likely that most/all of those games are crashing due to different errors that end up manifesting as an attempt to access invalid memory regions and each game will probably need to be investigated on a case-by-case basis (though with the two `Brother In Arms` games I'm optimistic that they may share source code and have the same underlying issue).

E.g., RtCW would've printed a similar `MMU Fault`, but the true cause was actually due to something far away from where the crash finally happened.
